File: src/JavaScript/ECv1Protocol.js
const fs = require('fs');
const zlib = require('zlib');

function readText(path) {
  if (path === '-' || path === undefined) {
    return fs.readFileSync(0, 'utf-8'); // stdin
  }
  return fs.readFileSync(path, 'utf-8');
}

function writeText(text, path) {
  if (!path || path === '-') {
    process.stdout.write(text);
    if (!text.endsWith('\n')) process.stdout.write('\n');
    return;
  }
  fs.writeFileSync(path, text);
}

function encodeMessage(inputFile, outputFile, transformChain = 'gz>b64', contentType = 'json') {
  const message = JSON.parse(readText(inputFile));
  const jsonBuffer = Buffer.from(JSON.stringify(message));

  let data = jsonBuffer;
  for (const transform of transformChain.split('>')) {
    if (transform === 'gz') {
      data = zlib.gzipSync(data);
    } else if (transform === 'b64') {
      data = Buffer.from(data.toString('base64'), 'utf8');
    } else if (transform === 'none' || transform === '') {
      continue;
    } else {
      throw new Error(`Unknown transform: ${transform}`);
    }
  }

  const ecv1Message = `EC v1\nt=${transformChain};ct=${contentType}\n${data.toString('utf8')}`;
  writeText(ecv1Message, outputFile);
}

function decodeMessage(encodedFile, outputFile) {
  const raw = readText(encodedFile);
  const lines = raw.replace(/\r\n/g, '\n').split('\n').filter(Boolean);
  if (lines.length < 3) throw new Error(`Invalid EC v1 message: expected 3 lines, got ${lines.length}`);
  if (lines[0] !== 'EC v1') throw new Error('Invalid EC v1 header');

  const { transformChain, contentType } = parseMeta(lines[1]);
  let data = Buffer.from(lines[2], 'utf8');

  // Apply transforms in reverse order to unwind encoding.
  for (const transform of transformChain.split('>').reverse()) {
    if (transform === 'b64') {
      data = Buffer.from(data.toString('utf8'), 'base64');
    } else if (transform === 'gz') {
      data = zlib.gunzipSync(data);
    } else if (transform === 'none' || transform === '') {
      continue;
    } else {
      throw new Error(`Unknown transform: ${transform}`);
    }
  }

  const text = data.toString('utf8');
  let parsed = text;
  if (contentType === 'json') {
    parsed = JSON.parse(text);
    writeText(JSON.stringify(parsed, null, 2), outputFile);
  } else {
    writeText(text, outputFile);
  }
}

function parseMeta(line) {
  const parts = line.split(';');
  let t = null, ct = 'json';
  for (const part of parts) {
    if (part.startsWith('t=')) t = part.slice(2);
    if (part.startsWith('ct=')) ct = part.slice(3);
  }
  if (!t) throw new Error('Transform chain (t=) not found');
  return { transformChain: t, contentType: ct || 'json' };
}

// Command-line interface
function usage() {
  console.log('Usage: node ECv1Protocol.js <encode|decode> <file|-> [outputFile|-]');
}

const args = process.argv.slice(2);
if (args.length < 2) {
  usage();
  process.exit(1);
}

const [mode, file, outputFile] = args;
try {
  if (mode === 'encode') {
    encodeMessage(file, outputFile);
  } else if (mode === 'decode') {
    decodeMessage(file, outputFile);
  } else {
    usage();
    process.exit(1);
  }
} catch (err) {
  console.error(`Error: ${err.message}`);
  process.exit(1);
}

File: src/CSharp/ECv1Protocol.cs
using System;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Text;
using System.Text.Json;

public static class ECv1Protocol
{
    // Encode a strongly typed message into EC v1 text.
    public static string Encode<T>(T message, string transformChain = "gz>b64", string contentType = "json")
    {
        var json = JsonSerializer.Serialize(message);
        var data = Encoding.UTF8.GetBytes(json);

        foreach (var transform in transformChain.Split('>'))
        {
            switch (transform)
            {
                case "gz":
                    using (var compressedStream = new MemoryStream())
                    {
                        using (var gzip = new GZipStream(compressedStream, CompressionLevel.Optimal))
                        {
                            gzip.Write(data, 0, data.Length);
                        }
                        data = compressedStream.ToArray();
                    }
                    break;
                case "b64":
                    data = Encoding.UTF8.GetBytes(Convert.ToBase64String(data));
                    break;
                case "none":
                    continue;
                default:
                    throw new InvalidOperationException($"Unknown transform: {transform}");
            }
        }

        var finalPayload = Encoding.UTF8.GetString(data);
        return $"EC v1\nt={transformChain};ct={contentType}\n{finalPayload}";
    }

    // Decode EC v1 text into a strongly typed object.
    public static T Decode<T>(string ecv1Text)
    {
        var lines = ecv1Text.Replace("\r\n", "\n").Split('\n').Where(l => !string.IsNullOrWhiteSpace(l)).ToArray();
        if (lines.Length < 3)
            throw new InvalidOperationException($"Invalid EC v1 message: expected 3 lines, got {lines.Length}.");
        if (lines[0] != "EC v1")
            throw new InvalidOperationException("Invalid EC v1 header.");

        var (transformChain, contentType) = ParseMetaLine(lines[1]);
        var payload = lines[2];

        var data = Encoding.UTF8.GetBytes(payload);
        data = ApplyTransformsInReverse(transformChain, data);

        var json = Encoding.UTF8.GetString(data);
        // contentType currently informational; future versions could branch on ct.
        return JsonSerializer.Deserialize<T>(json);
    }

    private static (string transformChain, string contentType) ParseMetaLine(string meta)
    {
        string t = null, ct = null;
        foreach (var part in meta.Split(';'))
        {
            if (part.StartsWith("t=")) t = part.Substring(2);
            if (part.StartsWith("ct=")) ct = part.Substring(3);
        }
        if (string.IsNullOrWhiteSpace(t))
            throw new InvalidOperationException("Transform chain (t=) not found.");
        if (string.IsNullOrWhiteSpace(ct))
            ct = "json";
        return (t, ct);
    }

    private static byte[] ApplyTransformsInReverse(string transformChain, byte[] data)
    {
        var chain = transformChain.Split('>', StringSplitOptions.RemoveEmptyEntries).Reverse();
        foreach (var transform in chain)
        {
            switch (transform)
            {
                case "b64":
                    data = Convert.FromBase64String(Encoding.UTF8.GetString(data));
                    break;
                case "gz":
                    using (var compressedStream = new MemoryStream(data))
                    using (var gzip = new GZipStream(compressedStream, CompressionMode.Decompress))
                    using (var resultStream = new MemoryStream())
                    {
                        gzip.CopyTo(resultStream);
                        data = resultStream.ToArray();
                    }
                    break;
                case "none":
                    continue;
                default:
                    throw new InvalidOperationException($"Unknown transform: {transform}");
            }
        }
        return data;
    }
}

// Minimal CLI for quick adoption: dotnet run -- encode input.json output.txt
public static class Program
{
    public static void Main(string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("Usage: dotnet run -- encode|decode <inputFile> [outputFile]");
            Environment.Exit(1);
        }

        var mode = args[0];
        var input = args[1];
        var output = args.Length > 2 ? args[2] : null;

        try
        {
            switch (mode)
            {
                case "encode":
                    var json = File.ReadAllText(input);
                    var obj = JsonSerializer.Deserialize<object>(json);
                    var encoded = ECv1Protocol.Encode(obj);
                    WriteOutput(encoded, output);
                    break;
                case "decode":
                    var text = File.ReadAllText(input);
                    var decoded = ECv1Protocol.Decode<object>(text);
                    var pretty = JsonSerializer.Serialize(decoded, new JsonSerializerOptions { WriteIndented = true });
                    WriteOutput(pretty, output);
                    break;
                default:
                    Console.WriteLine("Mode must be encode or decode.");
                    Environment.Exit(1);
                    break;
            }
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error: {ex.Message}");
            Environment.Exit(1);
        }
    }

    private static void WriteOutput(string content, string path)
    {
        if (string.IsNullOrEmpty(path))
        {
            Console.WriteLine(content);
        }
        else
        {
            File.WriteAllText(path, content);
        }
    }
}
